org.eclipse.jdt.core/compiler/org/eclipse/jdt/internal/compiler/lookup/TypeSystem.java

```diff
@@ -123,6 +126,11 @@ final int hash(TypeBinding b) {
 			public int hashCode() {
 				final int prime=31;
 				int hashCode = 1 + hash(this.type);
+				if (this.enclosingType != null && this.enclosingType.getClass() == ParameterizedTypeBinding.class) {
```

Comment:
Would "this.enclosingType != null && this.enclosingType.isParameterizedType()" cover it?  Or do you also need this for `RawTypeBinding`?

Reply:
So the base issue is that RawTypeBinding extends PTB and  raw instances can appear during URB resolution, so we cannot use "instanceof PTB".

And you're not wrong: isParameterizedType() would actually work here, but you have to hunt down all overloads to see this: TypeBinding defines it to returns false; it is overloaded to return true in PTB and then to return false again in RTB - and there are no other subclasses of PTB. 

Now here "exactly for PTB instances" is really meant so I think it is better to express this as direct as possible.

Reply:
Right, but isn't ".getClass() == ParameterizedTypeBinding.class" much more expensive than ".isParameterizedType()"?

Reply:
No, at least hotspot optimizes this, see https://github.com/openjdk/jdk/blob/ccc6e169188019dab008df5ffb11b4279e14f8fc/src/hotspot/share/opto/subnode.cpp#L1068